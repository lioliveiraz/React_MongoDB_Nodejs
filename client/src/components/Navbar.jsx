import React, { useState } from "react";
import { handlerLogOut } from "../store/actions/auth";
import { connect } from "react-redux";
import Breadcrumbs from "@material-ui/core/Breadcrumbs";
import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";
import { navigationLinks } from "../helpers/globalVar";
import BaseButton from "./Base/BaseButton";
import { useStyles } from "./../assets/css/Navigation/navigation";
import { useHistory } from "react-router-dom";

import SvgIcon from "@material-ui/icons/MoreVert";
import HomeIcon from "@material-ui/icons/Home";

function Navbar({ handlerLogOut, isAuthenticated, poolNotification }) {
  const { DEVELOPERS_PROFILES, MY_PROFILE, MY_WALL, SING_IN, SING_UP, LOGOUT } =
    navigationLinks;

  const classes = useStyles();
  const history = useHistory();

  return (
    <Breadcrumbs className={classes.root}>
      <Link className={classes.links} color="inherit" href="/">
        <HomeIcon color="secondary" />
      </Link>
      <Link className={classes.links} color="inherit" href="/developers">
        <Typography color="textPrimary">{DEVELOPERS_PROFILES}</Typography>
      </Link>
      {isAuthenticated ? (
        <Breadcrumbs className={classes.subnav}>
          <Link className={classes.links} color="inherit" href="/my-profile">
            <Typography color="textPrimary">{MY_PROFILE}</Typography>
          </Link>

          <Link
            className={classes.links}
            color="inherit"
            href="/my-profile/wall"
          >
            {MY_WALL}({poolNotification})
          </Link>

          <BaseButton
            color="secondary"
            value={LOGOUT}
            handleClick={handlerLogOut}
            className={classes.button}
          />
        </Breadcrumbs>
      ) : (
        <Breadcrumbs className={classes.subnav}>
          <Link className={classes.links} color="inherit" href="/signup">
            {SING_UP}
          </Link>
          <BaseButton
            value={SING_IN}
            color="secondary"
            handleClick={() => history.push("/login")}
            className={classes.button}
          />
        </Breadcrumbs>
      )}{" "}
    </Breadcrumbs>
  );
}
const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
  poolNotification: state.wall.notification,
});
export default connect(mapStateToProps, { handlerLogOut })(Navbar);
