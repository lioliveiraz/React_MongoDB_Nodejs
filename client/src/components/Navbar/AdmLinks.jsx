import React from "react";
import { connect } from "react-redux";
import { MenuItem } from "@material-ui/core";
import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";
import { navigationLinks } from "../../helpers/globalVar";

function AdmLinks({ isAdm, handleClose, classes }) {
  const { CATEGORIES } = navigationLinks;

  if (!isAdm) return null;

  return (
    <>
      <MenuItem onClick={handleClose}>
        <Link className={classes.links} color="inherit" href="/categories">
          <Typography color="textPrimary">{CATEGORIES}</Typography>
        </Link>
      </MenuItem>
    </>
  );
}
const mapStateToProps = (state) => {
  return {
    isAdm: state.auth.adm,
  };
};

export default connect(mapStateToProps)(AdmLinks);
