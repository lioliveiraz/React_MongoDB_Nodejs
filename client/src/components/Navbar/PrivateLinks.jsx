import React from "react";
import { connect } from "react-redux";
import { MenuItem } from "@material-ui/core";
import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";
import { navigationLinks } from "../../helpers/globalVar";

function PrivateLinks({ isAuthenticated, classes, handleClose }) {
  const { ADD_NEWTECH } = navigationLinks;

  if (!isAuthenticated) return null;

  return (
    <MenuItem onClick={handleClose}>
      <Link className={classes.links} color="inherit" href="/new-tech">
        <Typography color="textPrimary">{ADD_NEWTECH}</Typography>
      </Link>
    </MenuItem>
  );
}
const mapStateToProps = (state) => {
  return {
    isAuthenticated: state.auth.isAuthenticated,
  };
};

export default connect(mapStateToProps)(PrivateLinks);
