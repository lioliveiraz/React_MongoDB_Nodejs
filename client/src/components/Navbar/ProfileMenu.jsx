import React from "react";
import Menu from "@material-ui/core/Menu";
import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";
import { MenuItem } from "@material-ui/core";
import { navigationLinks } from "../../helpers/globalVar";

function ProfileMenu({ profileMenu, setProfileMenu, classes }) {
  const { MY_PROFILE, MY_WALL, EDIT_PROFILE } = navigationLinks;
  const handleCloseProfile = () => {
    setProfileMenu(null);
  };
  return (
    <Menu
      anchorEl={profileMenu}
      open={Boolean(profileMenu)}
      onClose={handleCloseProfile}
      className={classes.menu}
    >
      <MenuItem onClick={handleCloseProfile}>
        <Link className={classes.links} color="inherit" href="/my-profile">
          <Typography color="textPrimary">{MY_PROFILE}</Typography>
        </Link>
      </MenuItem>
      <MenuItem onClick={handleCloseProfile}>
        <Link className={classes.links} color="inherit" href="/edit-profile">
          <Typography color="textPrimary">{EDIT_PROFILE}</Typography>
        </Link>
      </MenuItem>
      <MenuItem onClick={handleCloseProfile}>
        <Link className={classes.links} color="inherit" href="/my-profile/wall">
          {MY_WALL}
        </Link>
      </MenuItem>
    </Menu>
  );
}

export default ProfileMenu;
