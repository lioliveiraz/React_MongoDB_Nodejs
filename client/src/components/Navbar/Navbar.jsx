import React, { useState } from "react";
import { handlerLogOut } from "../../store/actions/auth";
import { connect } from "react-redux";
import { useStyles } from "../../assets/css/Navigation/navigation";
import { useHistory } from "react-router-dom";

import { AppBar, MenuItem } from "@material-ui/core";
import { navigationLinks } from "../../helpers/globalVar";

import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";
import IconButton from "@material-ui/core/IconButton";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import Toolbar from "@material-ui/core/Toolbar";
import AccountCircle from "@material-ui/icons/AccountCircle";
import ExitToAppIcon from "@material-ui/icons/ExitToApp";
import Menu from "@material-ui/core/Menu";
import Button from "@material-ui/core/Button";
import ProfileMenu from "./ProfileMenu";
import AdmLinks from "./AdmLinks";
import PrivateLinks from "./PrivateLinks";

function Navbar({ handlerLogOut, isAuthenticated }) {
  const { DEVELOPERS_PROFILES, SING_IN, SING_UP } = navigationLinks;
  const [anchorEl, setAnchorEl] = useState(null);
  const [profileMenu, setProfileMenu] = useState(null);

  const classes = useStyles();
  const history = useHistory();

  const handleMainMenu = (e) => {
    setAnchorEl(e.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };
  const handleProfileMenu = (e) => {
    setProfileMenu(e.currentTarget);
  };

  return (
    <AppBar position="static" color="primary">
      <Toolbar className={classes.root}>
        <IconButton onClick={handleMainMenu} color="inherit">
          <MoreVertIcon />
        </IconButton>

        <Link color="inherit" href="/">
          <Typography variant="h4" color="inherit">
            NAME
          </Typography>
        </Link>

        <Menu
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={handleClose}
          className={classes.menu}
        >
          <MenuItem onClick={handleClose}>
            <Link className={classes.links} color="inherit" href="/developers">
              <Typography color="textPrimary">{DEVELOPERS_PROFILES}</Typography>
            </Link>
          </MenuItem>
          <PrivateLinks classes={classes} handleClose={handleClose} />
          <AdmLinks classes={classes} handleClose={handleClose} />

          {!isAuthenticated ? (
            <MenuItem onClick={handleClose}>
              <Link className={classes.links} color="inherit" href="/signup">
                {SING_UP}
              </Link>
            </MenuItem>
          ) : null}
        </Menu>

        {isAuthenticated ? (
          <div>
            <IconButton color="inherit" onClick={handleProfileMenu}>
              <AccountCircle />
            </IconButton>
            <IconButton color="inherit" onClick={handlerLogOut}>
              <ExitToAppIcon />
            </IconButton>
          </div>
        ) : (
          <Button
            variant="outlined"
            color="inherit"
            onClick={() => history.push("/login")}
            size="small"
          >
            <Typography variant="button">{SING_IN}</Typography>
          </Button>
        )}

        <ProfileMenu
          profileMenu={profileMenu}
          setProfileMenu={setProfileMenu}
          classes={classes}
        />
      </Toolbar>
    </AppBar>
  );
}
const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
  poolNotification: state.myWall.notification,
});
export default connect(mapStateToProps, { handlerLogOut })(Navbar);
