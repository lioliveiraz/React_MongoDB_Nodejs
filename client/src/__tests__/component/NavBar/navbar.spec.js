import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

import Navbar from "../../../components/Navbar/Navbar";
import { validateTruthiness } from "../../utils/testUtils";

describe("NavBar", () => {
  let wrapper, initialState, mockStore, store, historyMock;

  it("should match snapshot", () => {
    initialState = {
      auth: {
        isAuthenticated: true,
      },
    };
    mockStore = configureStore();
    store = mockStore(initialState);
    historyMock = { push: jest.fn() };
    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <Navbar historyMock={historyMock} />
        </Provider>
      </BrowserRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render correctly with user logged", async () => {
    initialState = {
      auth: {
        isAuthenticated: true,
      },
    };
    mockStore = configureStore();
    store = mockStore(initialState);
    historyMock = { push: jest.fn() };
    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <Navbar historyMock={historyMock} />
        </Provider>
      </BrowserRouter>
    );
    const profileMenu = wrapper.exists("[data-cy='profile-menu-toggle']");
    const logoutIcon = wrapper.exists("[data-cy='logout-icon']");
    const elementsArr = [profileMenu, logoutIcon];
    elementsArr.forEach((el) => {
      validateTruthiness(el);
    });

    wrapper.unmount();
  });

  it("the toggle buttons should work correctly", async () => {
    initialState = {
      auth: {
        isAuthenticated: true,
      },
    };
    mockStore = configureStore();
    store = mockStore(initialState);
    historyMock = { push: jest.fn() };
    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <Navbar historyMock={historyMock} />
        </Provider>
      </BrowserRouter>
    );
    const MenuButton = wrapper.find("[data-cy='navbar-menu-button']").at(0);
    await MenuButton.simulate("click");

    expect(wrapper.exists("ul")).toBeTruthy();

    await wrapper
      .find("[data-cy='profile-menu-toggle']")
      .at(0)
      .simulate("click");
    expect(wrapper.exists("ul")).toBeTruthy();
  });

  it("should render correctly with user not logged", () => {
    initialState = {
      auth: {
        isAuthenticated: false,
      },
    };
    mockStore = configureStore();
    store = mockStore(initialState);
    historyMock = { push: jest.fn() };
    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <Navbar historyMock={historyMock} />
        </Provider>
      </BrowserRouter>
    );
    const navBar = wrapper.exists('[data-cy="navbar"]');
    const MenuButton = wrapper.exists("[data-cy='navbar-menu-button']");
    const AppTittle = wrapper.exists("h4");
    const Menu = wrapper.exists('[data-cy="menu"]');
    const loginButton = wrapper.exists("[data-cy='login-button']");

    const elementsArr = [navBar, MenuButton, AppTittle, Menu, loginButton];
    elementsArr.forEach((el) => {
      validateTruthiness(el);
    });

    wrapper.unmount();
  });
});
