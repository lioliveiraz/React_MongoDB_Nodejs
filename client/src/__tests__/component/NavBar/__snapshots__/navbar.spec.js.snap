// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should match snapshot 1`] = `ReactWrapper {}`;

exports[`NavBar should match snapshot 2`] = `"<header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\" data-cy=\\"navbar\\"><div class=\\"MuiToolbar-root MuiToolbar-regular makeStyles-root-1 MuiToolbar-gutters\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" data-cy=\\"navbar-menu-button\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"/\\"><h4 class=\\"MuiTypography-root MuiTypography-h4 MuiTypography-colorInherit\\">COOL WALL</h4></a><div><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" data-cy=\\"profile-menu-toggle\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" data-cy=\\"logout-icon\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></header>"`;
